include/ginkgo/core/base/range_accessor_helper.hpp

```diff
@@ -0,0 +1,586 @@
+            ValueType, mask, set_bits_processed, stride_size, dim_idx + 1,
+            total_dim>::compute_idx(size, stride,
+                                    std::forward<Indices>(idxs)...);
+    }
+};
+
+// Specialization for the end of recursion: build_stride array from created
+// arguments
+template <typename ValueType, size_type mask, size_type set_bits_processed,
+          size_type stride_size, size_type total_dim>
+struct row_major_masked_helper_s<ValueType, mask, set_bits_processed,
+                                 stride_size, total_dim, total_dim, false> {
+    static_assert(set_bits_processed <= stride_size + 1,
+                  "The processed bits must be smaller than the total number of "
+                  "set bits!");
+    template <typename... Args>
+    static constexpr GKO_ATTRIBUTES GKO_INLINE
+        std::array<ValueType, stride_size>
+        build_stride(const dim<total_dim> &size, Args... args)
+    {
+        return {{std::forward<Args>(args)...}};
+    }
+    static constexpr GKO_ATTRIBUTES GKO_INLINE ValueType
+    mult_array(const dim<total_dim> &size)
+    {
+        return 1;
+    }
+
+    static constexpr GKO_ATTRIBUTES GKO_INLINE ValueType
+    compute_idx(const dim<total_dim> &size,
+                const std::array<ValueType, stride_size> &stride)
+    {
+        return 0;
+    }
+};
+
+
+}  // namespace detail
+
+
+/**
+ * Computes the memory index for the given indices considering the stride.
+ * Only indices are considered where the corresponding mask bit is set.
+ */
+template <typename ValueType, size_type mask, size_type stride_size,
+          size_type total_dim, typename... Indices>
+constexpr GKO_ATTRIBUTES GKO_INLINE auto compute_masked_index(
+    const dim<total_dim> &size,
+    const std::array<ValueType, stride_size> &stride, Indices &&... idxs)
+{
+    return detail::row_major_masked_helper_s<
+        ValueType, mask, 0, stride_size, 0,
+        total_dim>::compute_idx(size, stride, std::forward<Indices>(idxs)...);
+}
+
+
+/**
+ * Computes the default stride array from a size and a given mask which
+ * indicates which array indices to consider. It is assumed that there is no
+ * padding
+ */
+template <typename ValueType, size_type mask, size_type stride_size,
+          size_type total_dim>
+constexpr GKO_ATTRIBUTES GKO_INLINE auto compute_masked_stride_array(
+    const dim<total_dim> &size)
+{
+    return detail::row_major_masked_helper_s<ValueType, mask, 0, stride_size, 0,
+                                             total_dim>::build_stride(size);
+}
+
+
+/**
+ * Evaluates if all Args fulfill std::is_integral
+ */
+template <typename... Args>
+struct are_all_integral : public std::true_type {
+};
+
+template <typename First, typename... Args>
+struct are_all_integral<First, Args...>
+    : public std::conditional<std::is_integral<std::decay_t<First>>::value,
+                              are_all_integral<Args...>,
+                              std::false_type>::type {
+};
```

Comment:
Maybe this one can be moved to `types.hpp`? This seems like something which could be reused, and I'm not sure we would look for it here.